File: code/library/src/main/java/org/fbme/lib/iec61499/parser/ResourceTypeConverter.java
package org.fbme.lib.iec61499.parser;


import org.fbme.lib.common.Identifier;
import org.fbme.lib.iec61499.declarations.FBTypeReference;
import org.fbme.lib.iec61499.declarations.ResourceTypeDeclaration;
import org.jdom.Element;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public final class ResourceTypeConverter extends DeclarationConverterBase<ResourceTypeDeclaration> {

    public ResourceTypeConverter(ConverterArguments arguments) {
        super(arguments);
    }

    @Override
    protected ResourceTypeDeclaration extractDeclarationBody(@Nullable Identifier identifier) {
        ResourceTypeDeclaration rtd = myFactory.createResourceTypeDeclaration(identifier);

        ParameterDeclarationConverter.extractAll(this, rtd.getParameters());
        assert myElement != null;

        new FBNetworkConverter(with(myElement.getChild("FBNetwork")), rtd.getNetwork()).extractNetwork();

        List<Element> fbTypeNameElements = myElement.getChildren("FBTypeName");

        for (Element fbTypeNameElement : fbTypeNameElements) {
            FBTypeReference fbTypeReference = myFactory.createFBTypeReference();
            fbTypeReference.getFBType().setTargetName(fbTypeNameElement.getAttributeValue("Name"));
            rtd.getInstantiableFBTypes().add(fbTypeReference);
        }
        return rtd;
    }
}

File: code/library/src/main/kotlin/org/fbme/lib/iec61499/parser/ResourceTypeConverter.kt
package org.fbme.lib.iec61499.parser

import org.fbme.lib.common.Identifier
import org.fbme.lib.iec61499.declarations.ResourceTypeDeclaration

class ResourceTypeConverter(arguments: ConverterArguments) :
    DeclarationConverterBase<ResourceTypeDeclaration>(arguments) {
    override fun extractDeclarationBody(identifier: Identifier?): ResourceTypeDeclaration {
        val rtd = factory.createResourceTypeDeclaration(identifier)
        ParameterDeclarationConverter.extractAll(this, rtd.parameters)
        checkNotNull(element)
        FBNetworkConverter(with(element.getChild("FBNetwork")), rtd.network).extractNetwork()
        val fbTypeNameElements = element.getChildren("FBTypeName")
        for (fbTypeNameElement in fbTypeNameElements) {
            val fbTypeReference = factory.createFBTypeReference()
            fbTypeReference.fBType.setTargetName(fbTypeNameElement.getAttributeValue("Name"))
            rtd.instantiableFBTypes.add(fbTypeReference)
        }
        return rtd
    }
}
